applications/FluidDynamicsApplication/python_scripts/cfl_output_process.py

```diff
@@ -122,45 +143,35 @@ def _PrintToScreen(self, result_msg):
         KratosMultiphysics.Logger.PrintInfo(
             "CFLOutputProcess", "Current time: " + result_msg)
 
-    def _CalculateWithRespectToThreshold(self, x):
+    def _CalculateWithRespectToThreshold(self):
+        current_container = spatial_methods.NonHistorical.Elements.NormMethods
+
+        _, _, _, group_histogram, group_percentage_distribution, group_means, group_variances = current_container.Distribution(
+            self.model_part, KratosMultiphysics.CFL_NUMBER, "value", self.distribution_params)
 
-        y = [val for val in x if val < self.cfl_output_limit]
-        y1 = [val for val in x if val < 1.0]
         # % of element with cfl above threshold
-        how_many = ((len(x)-len(y))/len(x))*100
+        how_many = group_percentage_distribution[-1]*100.0
         # % of element with cfl above 1
-        how_many1 = ((len(x)-len(y1))/len(x))*100
+        how_many1 = (1.0 - group_percentage_distribution[0])*100.0
 
         # quantifying the mean and std for values below the threshold
-        y_mean = mean(y)
-        y_std = stdev(y)
+        total_elements_in_threshold_range = group_histogram[0] + group_histogram[1]
+        if (total_elements_in_threshold_range > 0):
+            y_mean = (group_means[0] * group_histogram[0] + group_means[1] * group_histogram[1]) / total_elements_in_threshold_range
+
+            threshold_sum_squared = (group_variances[0] + pow(group_means[0], 2)) * group_histogram[0] + (group_variances[1] + pow(group_means[1], 2)) * group_histogram[1]
+            y_std = sqrt((threshold_sum_squared  - total_elements_in_threshold_range * pow(y_mean, 2)) / (total_elements_in_threshold_range - 1.0))
+        else:
+            y_mean = 0.0
+            y_std = 0.0
 
         # qunatifying the global max
-        x_max = max(x)
+        # TODO: @Mate, where should we put the id of the element, where max is?
+        x_max, max_id = current_container.Max(self.model_part, KratosMultiphysics.CFL_NUMBER, "value")
         return [y_mean, y_std, x_max, how_many, how_many1]
 
     def _EvaluateCFL(self):
-
         if (self.model_part.ProcessInfo[KratosMultiphysics.DOMAIN_SIZE] == 2):
             KratosCFD.EstimateDtUtility2D.CalculateLocalCFL(self.model_part)
         else:
             KratosCFD.EstimateDtUtility3D.CalculateLocalCFL(self.model_part)
-
-        local_cfl = []
-        for elem in self.model_part.Elements:
-            local_cfl.append(elem.GetValue(KratosMultiphysics.CFL_NUMBER))
-
-        local_cfl = self.model_part.GetCommunicator().GetDataCommunicator().GathervDoubles(local_cfl, 0)
-
-        return local_cfl
-
-    def _SummarizeCFL(self, local_cfl):
-
-        global_cfl = []
-        for k in local_cfl:
-            global_cfl.extend(k)
-
-        cfl_mean, cfl_std, cfl_max, cfl_how_many, cfl_how_many1 = self._CalculateWithRespectToThreshold(
-            global_cfl)
-
-        return [cfl_mean, cfl_std, cfl_max, cfl_how_many, cfl_how_many1]
```

Comment:
Why is this piece of code removed?

Reply:
So, now everything can be done in C++ level, thus the performance boost. If the dependency is the case, then we can think about making a duplicate cfl_output_process in statistics application which is based on C++ level. But it will be again code duplication. I'm not sure how we can overcome that problem..:/

Reply:
OK. Now I see. Forget about my comment.